File: pkg/blunderdb/domain/xgid_contract_test.go
package domain

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

// xgidContractCase mirrors one entry of testdata/xgid_corpus.json. The corpus
// is shared with the GUI: this side decodes `xgid` and must produce exactly
// `position` and `pips` (and the board part of `xgidCanonical`); the GUI side
// (frontend/src/__tests__/xgidContract.test.js) starts from that same
// `position` and must re-encode it to `xgidCanonical` (generateXGID) and count
// `pips` (computePipCount). The GUI has no XGID parser any more (commit
// cd33de85), so `position` is the hand-off point between the two halves.
// The flat fields restate `position` for readability and are cross-checked
// against it. See the corpus _comment for conventions and the deliberately
// lossy re-encoding.
type xgidContractCase struct {
	Name          string   `json:"name"`
	XGID          string   `json:"xgid"`
	CubeOwner     int      `json:"cubeOwner"`
	CubeValueExp  int      `json:"cubeValueExp"`
	Dice          [2]int   `json:"dice"`
	PlayerOnRoll  int      `json:"playerOnRoll"`
	Score         [2]int   `json:"score"`
	HasJacoby     int      `json:"hasJacoby"`
	HasBeaver     int      `json:"hasBeaver"`
	Pips          [2]int   `json:"pips"`
	XGIDCanonical string   `json:"xgidCanonical"`
	Position      Position `json:"position"`
}

func TestDecodeXGIDContract(t *testing.T) {
	// domain test cwd is the package dir; the shared corpus lives at repo root.
	path := filepath.Join("..", "..", "..", "testdata", "xgid_corpus.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read corpus %s: %v", path, err)
	}
	var corpus struct {
		Cases []xgidContractCase `json:"cases"`
	}
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatalf("parse corpus: %v", err)
	}
	if len(corpus.Cases) == 0 {
		t.Fatal("corpus has no cases")
	}

	for _, c := range corpus.Cases {
		t.Run(c.Name, func(t *testing.T) {
			pos, err := DecodeXGID(c.XGID)
			if err != nil {
				t.Fatalf("DecodeXGID(%q): %v", c.XGID, err)
			}

			// The flat fields are what the corpus reader sees first; they
			// must agree with the decoder and with the `position` document.
			if pos.Cube.Owner != c.CubeOwner {
				t.Errorf("cubeOwner: got %d, want %d", pos.Cube.Owner, c.CubeOwner)
			}
			if pos.Cube.Value != c.CubeValueExp {
				t.Errorf("cubeValueExp: got %d, want %d", pos.Cube.Value, c.CubeValueExp)
			}
			if pos.Dice != c.Dice {
				t.Errorf("dice: got %v, want %v", pos.Dice, c.Dice)
			}
			if pos.PlayerOnRoll != c.PlayerOnRoll {
				t.Errorf("playerOnRoll: got %d, want %d", pos.PlayerOnRoll, c.PlayerOnRoll)
			}
			if pos.Score != c.Score {
				t.Errorf("score: got %v, want %v", pos.Score, c.Score)
			}
			if pos.HasJacoby != c.HasJacoby {
				t.Errorf("hasJacoby: got %d, want %d", pos.HasJacoby, c.HasJacoby)
			}
			if pos.HasBeaver != c.HasBeaver {
				t.Errorf("hasBeaver: got %d, want %d", pos.HasBeaver, c.HasBeaver)
			}

			// `position` is the exact document the GUI test starts from: the
			// decoder must reproduce it field for field (board, bearoff, cube,
			// dice, score, player on roll, decision type, jacoby, beaver).
			// ID and the provenance flags are not part of an XGID.
			want := c.Position
			want.ID, want.IndividuallyImported, want.Flagged = pos.ID, pos.IndividuallyImported, pos.Flagged
			if !reflect.DeepEqual(pos, want) {
				t.Errorf("position: decoded\n%+v\ndiffers from corpus\n%+v", pos, want)
			}

			p1, p2 := pos.ComputePipCounts()
			if got := [2]int{p1, p2}; got != c.Pips {
				t.Errorf("pips: got %v, want %v", got, c.Pips)
			}

			// The board is the only part of the XGID this package re-encodes;
			// the GUI test covers the full canonical string.
			board := strings.SplitN(c.XGIDCanonical, ":", 2)[0]
			if got := EncodeXGIDBoard(&pos); got != board {
				t.Errorf("EncodeXGIDBoard: got %q, want %q (board of xgidCanonical)", got, board)
			}
		})
	}
}

File: frontend/src/__tests__/xgidContract.test.js
/**
 * xgidContract.test.js
 *
 * GUI half of the XGID contract corpus (testdata/xgid_corpus.json), the other
 * half being pkg/blunderdb/domain/xgid_contract_test.go.
 *
 * The GUI no longer parses XGIDs itself (commit cd33de85: it parses via the
 * backend), so this side does not decode anything. It starts from `position`,
 * the exact document the Go decoder is asserted to produce for `xgid`, and
 * checks the two GUI functions that read such a position:
 *   - generateXGID(position) must equal `xgidCanonical` — the re-encoding
 *     under the GUI's own conventions (see the corpus _comment): Jacoby/Beaver
 *     round-trip losslessly since D.11/#211, match length is reconstructed as
 *     the larger away score (harmless — it still decodes to the same away
 *     scores), Crawford is inferred from an away score of 1, max cube is
 *     emitted as 0 (blunderDB has no capped-cube concept), and an offered cube
 *     re-encodes as dice 00;
 *   - computePipCount(position) must equal `pips`, the Go ComputePipCounts.
 * The flat fields (cubeOwner, dice, score, ...) restate `position` for the
 * reader; they are cross-checked so a corpus edit cannot leave them stale.
 */

import { describe, test, expect, vi } from 'vitest';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

// generateXGID is pure, but positionService.js pulls the Wails bindings and
// databaseService in at module load; stub them so the module imports under jsdom.
vi.mock('../../wailsjs/go/database/Database.js', () => ({}));
vi.mock('../services/databaseService.js', () => ({
    setStatusBarMessage: vi.fn(),
    warningMessageStore: { subscribe: vi.fn(), set: vi.fn(), update: vi.fn() }
}));

const { generateXGID } = await import('../services/positionService.js');
const { computePipCount } = await import('../utils/boardGeometry.js');

const __dirname = dirname(fileURLToPath(import.meta.url));
const corpus = JSON.parse(readFileSync(resolve(__dirname, '../../../testdata/xgid_corpus.json'), 'utf8'));

describe('XGID contract corpus (GUI generateXGID / computePipCount vs Go DecodeXGID)', () => {
    test('corpus is non-empty', () => {
        expect(corpus.cases.length).toBeGreaterThan(0);
    });

    for (const c of corpus.cases) {
        describe(c.name, () => {
            test('the flat fields restate position', () => {
                expect(c.position.cube.owner).toBe(c.cubeOwner);
                expect(c.position.cube.value).toBe(c.cubeValueExp);
                expect(c.position.dice).toEqual(c.dice);
                expect(c.position.player_on_roll).toBe(c.playerOnRoll);
                expect(c.position.score).toEqual(c.score);
                expect(c.position.has_jacoby).toBe(c.hasJacoby);
                expect(c.position.has_beaver).toBe(c.hasBeaver);
            });

            test('generateXGID re-encodes the decoded position to xgidCanonical', () => {
                expect(generateXGID(c.position)).toBe(c.xgidCanonical);
            });

            test('computePipCount agrees with Go ComputePipCounts', () => {
                const { pipCount1, pipCount2 } = computePipCount(c.position);
                expect([pipCount1, pipCount2]).toEqual(c.pips);
            });
        });
    }
});
